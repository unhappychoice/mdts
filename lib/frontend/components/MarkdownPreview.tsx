import React from 'react';
import { Box } from '@mui/material';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

interface MarkdownPreviewProps {
  content: string;
}

const MarkdownPreview: React.FC<MarkdownPreviewProps> = ({ content }) => {
  return (
    <Box>
      <ReactMarkdown remarkPlugins={[remarkGfm]}>{content}</ReactMarkdown>
    </Box>
  );
};

export default MarkdownPreview;
