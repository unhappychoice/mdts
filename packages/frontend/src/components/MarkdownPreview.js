"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const material_1 = require("@mui/material");
const react_markdown_1 = __importDefault(require("react-markdown"));
const react_syntax_highlighter_1 = require("react-syntax-highlighter");
const prism_1 = require("react-syntax-highlighter/dist/esm/styles/prism");
const rehype_raw_1 = __importDefault(require("rehype-raw"));
const remark_gfm_1 = __importDefault(require("remark-gfm"));
const remark_slug_1 = __importDefault(require("remark-slug"));
const MarkdownPreview = ({ content }) => {
    const theme = (0, material_1.useTheme)();
    return ((0, jsx_runtime_1.jsx)(material_1.Box, { className: ["markdown-body", theme.palette.mode === 'dark' ? 'dark' : 'light'].join(' '), sx: { py: 2, px: 0, fontSize: '0.9rem' }, children: (0, jsx_runtime_1.jsx)(react_markdown_1.default, { remarkPlugins: [remark_gfm_1.default, remark_slug_1.default], rehypePlugins: [rehype_raw_1.default], components: {
                code(_a) {
                    var { node, inline, className, children } = _a, props = __rest(_a, ["node", "inline", "className", "children"]);
                    const match = /language-(\w+)/.exec(className || '');
                    return !inline && match ? ((0, jsx_runtime_1.jsx)(react_syntax_highlighter_1.Prism, Object.assign({ style: theme.palette.mode === 'dark' ? prism_1.nightOwl : prism_1.prism, className: 'syntax-highlighter', customStyle: { margin: 0, background: 'transparent' }, showLineNumbers: true, language: match[1], PreTag: "div" }, props, { children: String(children).replace(/\n$/, '') }))) : ((0, jsx_runtime_1.jsx)("code", Object.assign({ className: className }, props, { children: children })));
                },
            }, children: content }) }));
};
exports.default = MarkdownPreview;
