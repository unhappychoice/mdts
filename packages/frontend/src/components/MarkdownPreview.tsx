import { Box, useTheme } from '@mui/material';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import ReactMarkdown from 'react-markdown';
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { nightOwl, prism } from 'react-syntax-highlighter/dist/esm/styles/prism';
import rehypeRaw from 'rehype-raw';
import remarkGfm from 'remark-gfm';
import remarkSlug from 'remark-slug';
import Mermaid from './Mermaid';

interface MarkdownPreviewProps {
  content: string;
  selectedFilePath: string | null;
}

const MarkdownPreview: React.FC<MarkdownPreviewProps> = ({ content, selectedFilePath }) => {
  const theme = useTheme();
  const navigate = useNavigate();

  const resolvePath = (href: string) => {
    if (!selectedFilePath || href.startsWith('http') || href.startsWith('//') || href.startsWith('/')) {
      return href;
    }
    const path = selectedFilePath.split('/');
    path.pop();
    return `/${path.join('/')}/${href}`.replace(/\/\.\//g, '/');
  };

  return (
    <Box className={["markdown-body", theme.palette.mode === 'dark' ? 'dark' : 'light'].join(' ')} sx={{ py: 2, px: 0, fontSize: '0.9rem' }}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm, remarkSlug]}
        rehypePlugins={[rehypeRaw]}
        components={{
          a: ({ href, children }) => {
            if (!href) {
              return <a>{children}</a>;
            }
            if (href.startsWith('#')) {
              return <a href={href}>{children}</a>;
            }
            const resolvedHref = resolvePath(href);
            if (resolvedHref.startsWith('http') || resolvedHref.startsWith('//')) {
              return <a href={resolvedHref} target="_blank" rel="noopener noreferrer">{children}</a>;
            }
            return <a href={resolvedHref} onClick={(e) => { e.preventDefault(); navigate(resolvedHref); }}>{children}</a>;
          },
          code({ node, inline, className, children, ...props }) {
            const match = /language-(\w+)/.exec(className || '');
            if (match && match[1] === 'mermaid') {
              return <Mermaid chart={String(children).replace(/\n$/, '')} />;
            }

            return !inline && match ? (
              <SyntaxHighlighter
                style={theme.palette.mode === 'dark' ? nightOwl : prism}
                className={'syntax-highlighter'}
                customStyle={{  margin: 0, background: 'transparent' }}
                showLineNumbers={true}
                language={match[1]}
                PreTag="div"
                {...props}
              >
                {String(children).replace(/\n$/, '')}
              </SyntaxHighlighter>
            ) : (
              <code className={className} {...props}>
                {children}
              </code>
            );
          },
        }}
      >
        {content}
      </ReactMarkdown>
    </Box>
  );
};

export default MarkdownPreview;
