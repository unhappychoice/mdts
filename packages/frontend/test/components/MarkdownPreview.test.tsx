import React from 'react';
import { render } from '@testing-library/react';
import MarkdownPreview from '../../src/components/MarkdownPreview';

describe('MarkdownPreview', () => {
  test('renders correctly', () => {
    const { asFragment } = render(<MarkdownPreview content="# Hello" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
