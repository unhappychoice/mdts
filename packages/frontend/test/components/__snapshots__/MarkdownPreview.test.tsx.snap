// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MarkdownPreview renders correctly 1`] = `
<DocumentFragment>
  <div
    class="markdown-body light MuiBox-root css-x7jyj4"
  >
    <div
      data-testid="mock-react-markdown"
    >
      # Hello
    </div>
  </div>
</DocumentFragment>
`;
